nexus/db-queries/src/db/datastore/snapshot_replacement.rs

```diff
@@ -0,0 +1,1428 @@
+                            record.replacement_state,
+                            record.operating_saga_id,
+                        )))
+                    }
+                }
+            },
+
+            Err(e) => Err(public_error_from_diesel(e, ErrorHandler::Server)),
+        }
+    }
+
+    /// Transition from Running to Complete, clearing the operating saga id and
+    /// removing the associated `volume_repair` record.
+    pub async fn set_snapshot_replacement_step_complete(
+        &self,
+        opctx: &OpContext,
+        snapshot_replacement_step_id: Uuid,
+        operating_saga_id: Uuid,
+        old_snapshot_volume_id: Uuid,
+    ) -> Result<(), Error> {
+        type TxnError = TransactionError<Error>;
+
+        self.pool_connection_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
+                use db::schema::volume_repair::dsl as volume_repair_dsl;
+
+                diesel::delete(
+                    volume_repair_dsl::volume_repair.filter(
+                        volume_repair_dsl::repair_id
+                            .eq(snapshot_replacement_step_id),
+                    ),
+                )
+                .execute_async(&conn)
+                .await?;
+
+                use db::schema::snapshot_replacement_step::dsl;
+                let result = diesel::update(dsl::snapshot_replacement_step)
+                    .filter(dsl::id.eq(snapshot_replacement_step_id))
+                    .filter(dsl::operating_saga_id.eq(operating_saga_id))
+                    .filter(dsl::old_snapshot_volume_id.is_null())
+                    .filter(
+                        dsl::replacement_state
+                            .eq(SnapshotReplacementStepState::Running),
+                    )
+                    .set((
+                        dsl::replacement_state
+                            .eq(SnapshotReplacementStepState::Complete),
+                        dsl::operating_saga_id.eq(Option::<Uuid>::None),
+                        dsl::old_snapshot_volume_id.eq(old_snapshot_volume_id),
+                    ))
+                    .check_if_exists::<SnapshotReplacementStep>(
+                        snapshot_replacement_step_id,
+                    )
+                    .execute_and_check(
+                        &*self.pool_connection_authorized(opctx).await?,
+                    )
+                    .await?;
+
+                match result.status {
+                    UpdateStatus::Updated => Ok(()),
+                    UpdateStatus::NotUpdatedButExists => {
+                        let record = result.found;
+
+                        if record.operating_saga_id == None
+                            && record.replacement_state
+                                == SnapshotReplacementStepState::Complete
+                        {
+                            Ok(())
+                        } else {
+                            Err(TxnError::CustomError(Error::conflict(
```

Comment:
Same question here as above about the error path. We have already deleted the volume
repair at this point, so what happens next? (Assuming this is true)

Reply:
This also executes in a transaction :)